### PR TITLE
Copy paste

### PR DESCRIPTION
**Pull request by karenzheung**
_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._

## Submitter:
JIRA 1618

Fixes the copy and paste bug with zero-width space on structured fields by eliminating the use of void nodes. 

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [ ] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [ ] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [x] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [x] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [x] Document any new APIs/extension points
- [x] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [x] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [x] Code is commented

**Tests**

- [x] Existing tests passed
- [x] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by jafeltra** [comment]
[JA3] If I highlight starting directly next to a shortcut (I think I am _not_ grabbing the zero-width-space) and copy and paste the text, it does not paste as structured data. Do we want it to?

[JA2] I'm getting a few warnings in the console about unused variables and an `expected !== but saw !=` warning.

[JA3] There seems to be a lot of whitespace changes that we can get rid of. For example, in the `by-key.js` file, we can probably get rid of the full diff we we delete the whitespace changes made. There are other lines that have extra spaces (like various lines in SingleHashtagKeywordStructuredFieldPlugin from lines 61 - 70). Ideally, I think new lines should not have any spaces or tabs in them. I can point out more specifics if you need them and if it's helpful.

**Comment by jafeltra** [comment]
[JA4] Using templates has a two bugs.
1) When you hover over an option in the picklist, the text displays twice in the note. 
2) Choosing fracture from the picklist options gives an error in the console:
`Uncaught TypeError: Cannot read property 'length' of null
    at FluxNotesEditor.resetShortcutData (FluxNotesEditor.jsx:586)`
[JA5] I can get to a state when I lose my cursor. I'm not sure if this is related to changing how we render the shortcuts, but I can't seem to get it to happen on fluxnotes.org. If I insert condition as the first thing in the note, then arrow to the left, I lose the cursor. On fluxnotes.org, it appears on the left side of the shortcut after 2 arrow clicks. I also can't ever get the cursor to come back after hitting right arrow. When I am in this state and can't see the cursor, I also can't type anything.

**Comment by mgramigna** [comment]
[MG1] These are more questions than change requests, I am just looking for more insight so I understand the code changes more.

1) What is the reasoning for the Slate changes in `apply-operation.js` and `operation.js`?
2) What is the reasoning for the changes in the `resetShortcutData` function in `FluxNotesEditor.jsx`?

I'm just having a hard time following, and I think we should be careful about making changes to Slate's code so I just want to understand the reasoning.

**Comment by karenzheung** [comment]
Fixed all the JAs. There is an issue where paste doesn't work if the zero width space is not selected. 
Regarding MG1, I made those changes because insertText and removeText expects a text node, so when a block or inline is passed I get an error. I found this issue regarding that: 
https://github.com/ianstormtaylor/slate/issues/775

The changes in resetShortcutData were made because when inserting from the picklist, it didn't properly update the shortcut with the new selected value since the rendering of shortcuts were changed. If I selected @condition, then selected Fracture from the picklist the editor would show @conditionFracture vs Fracture.

**Comment by mgramigna** [comment]
[MG2] The arrow key navigation works for hitting the left arrow, but hitting the right arrow at the beginning of an inserter causes it to jump to the end when we expect it to go character by character (works fine for hashtags, just need to keep the old arrow key functionality in for inserters)

Perhaps we should add a case to the if statement for the arrow keys to make sure that the shortcut we care about is not an Inserter before doing the custom arrow key navigation?

**Comment by karenzheung** [comment]
Fixed MG2

File: src/notes/SingleHashtagKeywordStructuredFieldPlugin.jsx [hunk 1]
```diff
@@ -61,8 +62,8 @@ function SingleHashtagKeywordStructuredFieldPlugin(opts) {
                     let keywordRange;
                     if (curNode.nodes) {
                         for (const childNode of curNode.nodes) {
-                            keywordRange = getRangeForKeyword(childNode, keywordText);
-                            if (keywordRange) break;
+                                keywordRange = getRangeForKeyword(childNode, keywordText);
```
Comment (mgramigna) [inline src/notes/SingleHashtagKeywordStructuredFieldPlugin.jsx]: Unnecessary whitespace (lines 65-66)

File: src/notes/StructuredFieldPlugin.jsx [hunk 2]
```diff
@@ -51,13 +51,43 @@ function StructuredFieldPlugin(opts) {
...
+            if (previousNode.type === 'structured_field'
+                && !(previousNode.data.get('shortcut') instanceof InsertValue)
+                && state.selection.anchorOffset === 0 && state.selection.isCollapsed) {
+                let transform = state.transform();
+                transform = transform.removeNodeByKey(previousNode.key);
+                let newState = transform.apply();
+                return newState;
+            }
+        } 
+        else if (e.keyCode === 37 && previousNode){
```
Comment (mgramigna) [inline src/notes/StructuredFieldPlugin.jsx]: Unnecessary single space (lines 73 and 81). Should probably just move the `else if (...)` up a line to be on the same line as the bracket to stay consistent with our JS style

Comment (mgramigna) [reply src/notes/StructuredFieldPlugin.jsx]: Also need a space on line 74 before the bracket

File: src/notes/StructuredFieldPlugin.jsx [hunk 3]
```diff
@@ -530,6 +560,7 @@ function StructuredFieldPlugin(opts) {
         utils: {
             //isSelectionInStructuredField
             convertSlateNodesToText: convertSlateNodesToText,
+            createStructuredField: createStructuredField.bind(null, opts)
```
Comment (mgramigna) [inline src/notes/StructuredFieldPlugin.jsx]: Not sure we need this line at all

File: src/shortcuts/InsertValue.jsx [hunk 4]
```diff
@@ -190,6 +190,7 @@ export default class InsertValue extends Shortcut {
     }
 
     getText() {
+
```
Comment (mgramigna) [inline src/shortcuts/InsertValue.jsx]: this can be removed

**Comment by mgramigna** [comment]
[MG4] We should be updating all the schemas (except placeholders) to reflect the new non-void hashtags.

i.e. update the schema for `structured_field_selected_search_result` and `structured_field_search_result` to also get rid of `shortcut.getText()` in the rendering, similar to what was done for `structured_field` and `bolded_structured_field`

**Comment by karenzheung** [comment]
Fixed MG 3 and MG4

File: src/notes/StructuredFieldPlugin.jsx (hunk 3 above)
Comment (mgramigna) [reply src/notes/StructuredFieldPlugin.jsx]: Not sure if the change for this got pushed, or do you still think we need this line?

**Comment by mgramigna** [comment]
I think this looks good. I will approve pending follow up on [this comment](https://github.com/FluxNotes/flux/pull/566#discussion_r262658120)

**Comment by jafeltra** [comment]
This looks good to me! All the previous JA's are resolved, pending the one last comment from Matt. I did notice one new thing that I think is related to the changes in SingleHashtagKeywordStructuredFieldPlugin.js:

[JA6]: If you type "#disease status #toxicity nausea disease " the first disease word gets captured as the structured phrase after you hit space and the second disease does not. I know this is a bit of a strange use case, but since we have a fix for it you type #disease status after #toxicity, I feel like we should support this case too.

**Comment by karenzheung** [comment]
Fixed JA6

File: src/notes/StructuredFieldPlugin.jsx (hunk 3 above)
Comment (karenzheung) [reply src/notes/StructuredFieldPlugin.jsx]: I forgot this push this. pushed now

**Comment by jafeltra** [comment]
JA6 looks good to me! I haven't noticed anything new so everything looks set to me! Nice job!!
